src/Ens.ts

```diff
@@ -0,0 +1,638 @@
+import {ResolutionError, ResolutionErrorCode} from './errors/resolutionError';
+import EthereumContract from './contracts/EthereumContract';
+import EnsNetworkMap from 'ethereum-ens-network-map';
+import {
+  EnsSource,
+  Locations,
+  NamingServiceName,
+  Provider,
+  TokenUriMetadata,
+  BlockchainType,
+  DnsRecordType,
+  DnsRecord,
+} from './types/publicTypes';
+import {EthereumNetworksInverted, isNullAddress} from './utils';
+import FetchProvider from './FetchProvider';
+import {eip137Childhash, eip137Namehash, labelNameHash} from './utils/namehash';
+import {NamingService} from './NamingService';
+import ConfigurationError, {
+  ConfigurationErrorCode,
+} from './errors/configurationError';
+import {EthereumNetworks} from './utils';
+import {requireOrFail} from './utils/requireOrFail';
+import ensConfig from '../src/config/ens-config.json';
+import Networking from './utils/Networking';
+
+/**
+ * @internal
+ */
+export default class Ens extends NamingService {
+  readonly name = NamingServiceName.ENS;
+  readonly network: number;
+  readonly networkName: string;
+  readonly url: string;
+  readonly provider: Provider;
+  readonly registryContract: EthereumContract;
+  readonly nameWrapperContract: EthereumContract;
+  readonly baseRegistrarContract: EthereumContract;
+  readonly proxyServiceApiKey: string | undefined;
+
+  constructor(source?: EnsSource) {
+    super();
+    let finalSource: EnsSource = {url: '', network: 'mainnet'};
+    if (source) {
+      finalSource = this.checkNetworkConfig(source);
+    }
+    this.network = EthereumNetworks[finalSource.network];
+    this.networkName = finalSource.network;
+    this.url = finalSource['url'];
+    this.provider =
+      finalSource['provider'] ||
+      FetchProvider.factory(NamingServiceName.ENS, this.url);
+    this.proxyServiceApiKey = finalSource['proxyServiceApiKey'];
+
+    const registryAddress =
+      finalSource['registryAddress'] || EnsNetworkMap[this.network];
+    this.registryContract = new EthereumContract(
+      ensInterface,
+      registryAddress,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+
+    const nameWrapperAddress = this.determineNameWrapperAddress(this.network);
+    this.nameWrapperContract = new EthereumContract(
+      nameWrapperInterface,
+      nameWrapperAddress,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+
+    const baseRegistrarAddress = this.determineBaseRegistrarAddress(
+      this.network,
+    );
+    this.baseRegistrarContract = new EthereumContract(
+      baseRegistrarInterface,
+      baseRegistrarAddress,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+  }
+
+  static async autoNetwork(
+    config: {url: string} | {provider: Provider},
+  ): Promise<Ens> {
+    let provider: Provider;
+
+    if (hasProvider(config)) {
+      provider = config.provider;
+    } else {
+      if (!config.url) {
+        throw new ConfigurationError(ConfigurationErrorCode.UnspecifiedUrl, {
```

Comment:
```suggestion
        throw new ConfigurationError(ConfigurationErrorCode.EnnSpecifiedUrl, {
```

Reply:
What does this error mean? EnsSpecifiedUrl?

Reply:
my bads, closing the comment